Feedback by Hannes (#82)
oauthstuff/draft-oauth-rar: main.md

The following example shows how an implementation could utilize the namespace `https://scheme.example.org/` to ensure collision-resistant type values.

```JSON
{
   "type": "https://scheme.example.org/files",
   "locations": [
      "https://example.com/files"
   ],
   "permissions": [
      {
         "path": "/myfiles/A",
         "access": [
            "read"
         ]
      },
      {
         "path": "/myfiles/A/X",
         "access": [
            "read",
            "write"
         ]
      }
   ]
}
```
Figure: Example for authorization details with an URL as type identifier.

## Common data fields {#common_data_fields}

This specification defines a set of common data fields that are designed to be usable across different types of APIs. These data fields MAY be combined in different ways depending on the needs of the API. All data fields are OPTIONAL for use by a given API definition. The allowable values of all fields are determined by the API being protected.

`locations`:
:   An array of strings representing the location of the resource or resource server. These strings are typically URIs identifying the

>     location of the RS. The field `locations` allows a client to specify where it intends to use a certain authorization, i.e., it is  possible to unambiguously assign permissions to resource servers. In situations with multiple resource servers, this prevents unintended client authorizations (e.g. a `read` scope value potentially applicable for an email as well as a cloud service) through audience restriction.

Reviewer: I would rather see this as a security consideration than in the definition here, but keep a forward reference. So extract the explanation text and leave something like this:

```suggestion
    location of the RS. This field can allow a client to specify a particular RS, as discussed in {{security-considerations-location-field-values}}.
```

Author: adopted your proposal

   }
]
```
Figure: URL decoded authorization details.

## Relationship to "scope" parameter {#scope}

`authorization_details` and `scope` can be used in the same authorization request for carrying independent authorization requirements. 

> Note: Combined use of `authorization_details` and `scope` is supported by this specification to allow existing OAuth-based applications to incrementally migration towards `authorization_details`. It is RECOMMENDED that a given API uses only one form of requirement specification. 

Reviewer: Would suggest rewording this so it's not an aside:

```suggestion
Combined use of `authorization_details` and `scope` is supported by this specification in part to allow existing OAuth-based applications to incrementally migrate towards using `authorization_details` exclusively. It is RECOMMENDED that a given API use only one form of requirement specification. 
```

# Authorization Error Response {#authz_details_error}

> The AS MUST refuse to process any unknown authorization details type or authorization details not conforming to the respective type definition. If any of the objects in `authorization_details` contains an unknown authorization details type or an object of known type but containing unknown fields or fields of the wrong type or fields 
> with invalid values or if required fields are missing, the AS MUST abort processing and respond with an error `invalid_authorization_details` to the client. 

Reviewer: It's more verbose but I think this should be spelled out:

```suggestion
The AS MUST refuse to process any unknown authorization details type or authorization details not conforming to the respective type definition. The AS MUST abort processing and respond with an error `invalid_authorization_details` to the client if any of the following are true of any of the objects in `authorization_details` structure:

- Contains an unknown authorization details type value,
- An object of known type but containing unknown fields,
- Contains fields of the wrong type for the authorization details type,
- Contains fields with invalid values for the authorization details type, or
- Missing required fields for the authorization details type.
```

   }
}
```
Figure: Example for authorization details in JWT-based access token.

> In this case, the AS added the following example claims:

Reviewer: ```suggestion
In this case, the AS added the following example claims to the JWT-based access token:
```

* `sub`: conveys the user on which behalf the client is asking for payment initation
* `txn`: transaction id used to trace the transaction across the services of provider `example.com`
* `debtorAccount`: API-specific field containing the debtor account. In the example, this account was not passed in the authorization details but selected by the user during the authorization process. The field `user_role` conveys the role the user has with respect to this particular account. In this case, they is the owner. This data is used for access control at the payment API (the RS).

## Token Introspection 

In the case of opaque access tokens, the data provided to a certain RS is determined using the RS's identifier with the AS (see [@I-D.ietf-oauth-jwt-introspection-response], section 3). 

> The token introspection response provides the RS with the authorization details applicable to it as a top-level JSON field along with the claims the RS requires for request processing. 

Reviewer: I believe "element" is actually correct here since we're talking about JSON itself

```suggestion
The token introspection response provides the RS with the authorization details applicable to it as a top-level JSON element along with the claims the RS requires for request processing. 
```

## OpenID Connect {#openid}

OpenID Connect [@OIDC] specifies the JSON-based `claims` request parameter that can be used to specify the claims a client (acting as OpenID Connect Relying Party) wishes to receive in a fine-grained and privacy-preserving way as well as assign those claims to certain delivery mechanisms, i.e. ID Token or userinfo response. 

The combination of the scope value `openid` and the additional parameter `claims` can be used beside `authorization_details` in the same way as every non-OIDC scope value. 

> Alternatively, there could be an authorization details type for OpenID Connect. This section gives an example of what such an authorization details type could look like.

Reviewer: ```suggestion
Alternatively, there could be an authorization details type for OpenID Connect. This section gives an example of what such an authorization details type could look like, but defining such a type is outside the scope of this specification.
```

   [[ To be removed from the final specification ]]

   -09

>    * Incorporated feedback by Hannes

Reviewer 2: ```suggestion
   * Incorporated feedback by Hannes as document shepherd
```

   -08

   * formatting in authorization details type section 

>    * added example for `privileges` common data field 

Reviewer: I wouldn't change the changelog (this is a nit)

```suggestion
   * added example for `privileges` common data element
```

   -04 

>    * restructured specification for better readability

Reviewer: I wouldn't change the changelog (this is a nit)

```suggestion
   * restructured draft for better readability
```
